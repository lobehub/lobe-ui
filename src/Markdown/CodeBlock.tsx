import Pre, { PreMermaid, PreSingleLine } from '@lobehub/ui/mdx/mdxComponents/Pre';
import { FC } from 'react';

import { type HighlighterProps } from '@/Highlighter';
import { type MermaidProps } from '@/Mermaid';
import { FALLBACK_LANG } from '@/hooks/useHighlight';

const countLines = (str: string): number => {
  const regex = /\n/g;
  const matches = str.match(regex);
  return matches ? matches.length : 1;
};

export const useCode = (raw: any) => {
  if (!raw) return;

  const { children, className } = raw.props;

  if (!children) return;

  const content = Array.isArray(children) ? (children[0] as string) : children;

  const lang = className?.replace('language-', '') || FALLBACK_LANG;

  const isSingleLine = countLines(content) <= 1 && content.length <= 32;

  return {
    content,
    isSingleLine,
    lang,
  };
};

interface CodeBlockProps {
  children: any;
  enableMermaid?: boolean;
  fullFeatured?: boolean;
  highlight?: HighlighterProps;
  mermaid?: MermaidProps;
}

const CodeBlock: FC<CodeBlockProps> = ({
  fullFeatured,
  enableMermaid,
  highlight,
  mermaid,
  children,
  ...rest
}) => {
  const code = useCode(children);

  if (!code) return;

  if (enableMermaid && code.lang === 'mermaid')
    return (
      <PreMermaid fullFeatured={fullFeatured} {...mermaid} {...rest}>
        {code.content}
      </PreMermaid>
    );

  if (!highlight && code.isSingleLine)
    return <PreSingleLine language={code.lang}>{code.content}</PreSingleLine>;

  return (
    <Pre fullFeatured={fullFeatured} language={code.lang} {...highlight} {...rest}>
      {code.content}
    </Pre>
  );
};

export const CodeLite: FC<CodeBlockProps> = (props) => {
  return <CodeBlock {...props} />;
};

export const CodeFullFeatured: FC<CodeBlockProps> = (props) => {
  return <CodeBlock fullFeatured {...props} />;
};
