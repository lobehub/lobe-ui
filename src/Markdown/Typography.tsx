'use client';

import { memo } from 'react';

import { useStyles } from './markdown.style';
import type { TypographyProps } from './type';

const Typography = memo<TypographyProps>(
  ({ ref, children, className, fontSize, headerMultiple, marginMultiple, lineHeight, ...rest }) => {
    const { cx, styles } = useStyles({ fontSize, headerMultiple, lineHeight, marginMultiple });
    return (
      <article className={cx(styles.root, styles.variant, className)} ref={ref} {...rest}>
        {children}
      </article>
    );
  },
);

Typography.displayName = 'Typography';

export default Typography;
