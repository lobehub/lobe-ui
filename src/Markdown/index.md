---
nav: Components
group: Data Display
title: Markdown
description: Markdown is a React component used to render markdown text. It supports various markdown syntax such as headings, lists, links, images, code blocks and more. It is commonly used in documentation, blogs, and other text-heavy applications.
---

## Default

<code src="./demos/index.tsx" nopadding></code>

## Chat Mode

<code src="./demos/chat.tsx" nopadding></code>

## Shiki Transformers

<code src="./demos/transformer.tsx" nopadding></code>

## Custom Markdown Components

<code src="./demos/customComponents.tsx" nopadding></code>

## Custom Highlight

<code src="./demos/customHighlight.tsx" nopadding></code>

## Think

<code src="./demos/thinking/index.tsx" nopadding></code>

## Custom Plugins

<code src="./demos/customPlugins/index.tsx" nopadding></code>

## APIs

<API></API>
