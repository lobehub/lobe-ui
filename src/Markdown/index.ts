export { default } from './Markdown';
export type * from './type';
export { default as Typography } from './Typography';
