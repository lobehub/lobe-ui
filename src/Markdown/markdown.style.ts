import { createStyles } from 'antd-style';
import { rgba } from 'polished';

const IGNORE_CLASSNAME = '.ignore-markdown-style';

export const useStyles = createStyles(
  (
    { cx, token, isDarkMode, css },
    {
      fontSize = 16,
      headerMultiple = 1,
      marginMultiple = 1.5,
      lineHeight = 1.8,
    }: { fontSize?: number; headerMultiple?: number; lineHeight?: number; marginMultiple?: number },
  ) => {
    const __root = css`
      --lobe-markdown-font-size: ${fontSize}px;
      --lobe-markdown-header-multiple: ${headerMultiple};
      --lobe-markdown-margin-multiple: ${marginMultiple};
      --lobe-markdown-line-height: ${lineHeight};
      --lobe-markdown-border-radius: ${token.borderRadiusLG};
      --lobe-markdown-border-color: ${isDarkMode
        ? token.colorBorderSecondary
        : rgba(token.colorBorderSecondary, 0.5)};

      position: relative;

      width: 100%;
      max-width: 100%;
      padding-inline: 1px;

      font-size: var(--lobe-markdown-font-size);
      line-height: var(--lobe-markdown-line-height);
      word-break: break-word;

      ${IGNORE_CLASSNAME} {
        font-size: 14px;
        line-height: 1.5;
      }
    `;
    const a = css`
      a:not(${IGNORE_CLASSNAME} a) {
        color: ${token.colorInfoText};

        &:hover {
          color: ${token.colorInfoHover};
        }
      }
    `;
    const blockquote = css`
      blockquote:not(${IGNORE_CLASSNAME} blockquote) {
        margin-block: calc(var(--lobe-markdown-margin-multiple) * 1em);
        margin-inline: 0;
        padding-block: 0;
        padding-inline: 1em;

        color: ${token.colorTextSecondary};

        border-inline-start: solid 4px ${token.colorBorder};
      }
    `;
    const code = css`
      code:not(${IGNORE_CLASSNAME} code) {
        &:not(:has(span)) {
          display: inline;

          margin-inline: 0.25em;
          padding-block: 0.2em;
          padding-inline: 0.4em;

          font-family: ${token.fontFamilyCode};
          font-size: 0.875em;
          line-height: 1;
          word-break: break-word;
          white-space: break-spaces;

          background: ${token.colorFillSecondary};
          border: 1px solid var(--lobe-markdown-border-color);
          border-radius: 0.25em;
        }
      }
    `;
    const details = css`
      details:not(${IGNORE_CLASSNAME} details) {
        margin-block: calc(var(--lobe-markdown-margin-multiple) * 1em);
        padding-block: 0.75em;
        padding-inline: 1em;

        background: ${token.colorFillTertiary};
        border-radius: calc(var(--lobe-markdown-border-radius) * 1px);
        box-shadow: 0 0 0 1px var(--lobe-markdown-border-color);

        summary {
          cursor: pointer;
          display: flex;
          align-items: center;
          list-style: none;

          &::before {
            content: '';

            position: absolute;
            inset-inline-end: 1.25em;
            transform: rotateZ(-45deg);

            display: block;

            width: 0.4em;
            height: 0.4em;

            font-family: ${token.fontFamily};

            border-block-end: 1.5px solid ${token.colorTextSecondary};
            border-inline-end: 1.5px solid ${token.colorTextSecondary};

            transition: transform 200ms ${token.motionEaseOut};
          }
        }

        &[open] {
          summary {
            padding-block-end: 0.75em;
            border-block-end: 1px dashed ${token.colorBorder};

            &::before {
              transform: rotateZ(45deg);
            }
          }
        }
      }
    `;
    const header = css`
      h1:not(${IGNORE_CLASSNAME} h1),
      h2:not(${IGNORE_CLASSNAME} h2),
      h3:not(${IGNORE_CLASSNAME} h3),
      h4:not(${IGNORE_CLASSNAME} h4),
      h5:not(${IGNORE_CLASSNAME} h5),
      h6:not(${IGNORE_CLASSNAME} h6) {
        margin-block: max(
          calc(var(--lobe-markdown-header-multiple) * var(--lobe-markdown-margin-multiple) * 0.4em),
          var(--lobe-markdown-font-size)
        );
        font-weight: bold;
        line-height: 1.25;
      }

      h1:not(${IGNORE_CLASSNAME} h1) {
        font-size: calc(
          var(--lobe-markdown-font-size) * (1 + 1.5 * var(--lobe-markdown-header-multiple))
        );
      }

      h2:not(${IGNORE_CLASSNAME} h2) {
        font-size: calc(
          var(--lobe-markdown-font-size) * (1 + var(--lobe-markdown-header-multiple))
        );
      }

      h3:not(${IGNORE_CLASSNAME} h3) {
        font-size: calc(
          var(--lobe-markdown-font-size) * (1 + 0.5 * var(--lobe-markdown-header-multiple))
        );
      }

      h4:not(${IGNORE_CLASSNAME} h4) {
        font-size: calc(
          var(--lobe-markdown-font-size) * (1 + 0.25 * var(--lobe-markdown-header-multiple))
        );
      }

      h5:not(${IGNORE_CLASSNAME} h5),
      h6:not(${IGNORE_CLASSNAME} h6) {
        font-size: calc(var(--lobe-markdown-font-size) * 1);
      }
    `;
    const hr = css`
      hr:not(${IGNORE_CLASSNAME} hr) {
        width: 100%;
        margin-block: calc(var(--lobe-markdown-margin-multiple) * 1.5em);

        border-color: ${token.colorBorder};
        border-style: dashed;
        border-width: 1px;
        border-block-start: none;
        border-inline-start: none;
        border-inline-end: none;
      }
    `;
    const img = css`
      img:not(${IGNORE_CLASSNAME} img) {
        max-width: 100%;
      }

      > img,
      > p > img {
        margin-block: calc(var(--lobe-markdown-margin-multiple) * 1em);
        border-radius: calc(var(--lobe-markdown-border-radius) * 1px);
        box-shadow: 0 0 0 1px var(--lobe-markdown-border-color);
      }
    `;
    const kbd = css`
      kbd:not(${IGNORE_CLASSNAME} kbd) {
        cursor: default;

        display: inline-block;

        min-width: 1em;
        margin-inline: 0.25em;
        padding-block: 0.2em;
        padding-inline: 0.4em;

        font-family: ${token.fontFamily};
        font-size: 0.875em;
        font-weight: 500;
        line-height: 1;
        text-align: center;

        background: ${token.colorBgLayout};
        border: 1px solid ${token.colorBorderSecondary};
        border-radius: 0.25em;
      }
    `;
    const list = css`
      li:not(${IGNORE_CLASSNAME} li) {
        margin-block: calc(var(--lobe-markdown-margin-multiple) * 0.33em);

        p {
          display: inline;
        }
      }

      ul:not(${IGNORE_CLASSNAME} ul),
      ol:not(${IGNORE_CLASSNAME} ol) {
        margin-block: calc(var(--lobe-markdown-margin-multiple) * 1em);
        margin-inline-start: 1em;
        padding-inline-start: 0;
        list-style-position: outside;

        > ul,
        > ol {
          margin-block: 0;
        }

        > li {
          margin-inline-start: 1em;
        }
      }

      ol:not(${IGNORE_CLASSNAME} ol) {
        list-style: auto;
      }

      ul:not(${IGNORE_CLASSNAME} ul) {
        list-style-type: none;

        > li {
          &::before {
            content: '-';
            display: inline-block;
            margin-inline: -1em 0.5em;
            opacity: 0.5;
          }
        }
      }
    `;
    const p = css`
      p:not(${IGNORE_CLASSNAME} kbd) {
        margin-block: 4px;
        line-height: var(--lobe-markdown-line-height);
        letter-spacing: 0.02em;

        &:not(:first-child) {
          margin-block-start: calc(var(--lobe-markdown-margin-multiple) * 1em);
        }

        &:not(:last-child) {
          margin-block-end: calc(var(--lobe-markdown-margin-multiple) * 1em);
        }
      }
    `;
    const pre = css`
      pre:not(${IGNORE_CLASSNAME} pre),
      [data-code-type='highlighter'] {
        white-space: break-spaces;
        border: none;

        > code {
          padding: 0 !important;

          font-family: ${token.fontFamilyCode};
          font-size: 0.875em;
          line-height: 1.6;

          border: none !important;
        }
      }
    `;
    const strong = css`
      strong:not(${IGNORE_CLASSNAME} strong) {
        font-weight: 600;
      }
    `;
    const svg = css`
      svg:not(${IGNORE_CLASSNAME} svg) {
        line-height: 1;
      }
    `;
    const table = css`
      table:not(${IGNORE_CLASSNAME} table) {
        unicode-bidi: isolate;
        overflow: auto hidden;
        display: block;
        border-spacing: 0;
        border-collapse: collapse;

        box-sizing: border-box;
        width: max-content;
        max-width: 100%;
        margin-block: calc(var(--lobe-markdown-margin-multiple) * 1em);

        text-align: start;
        text-indent: initial;
        text-wrap: pretty;
        word-break: auto-phrase;
        overflow-wrap: break-word;

        background: ${token.colorFillQuaternary};
        border-radius: calc(var(--lobe-markdown-border-radius) * 1px);
        box-shadow: 0 0 0 1px var(--lobe-markdown-border-color);

        code {
          word-break: break-word;
        }

        thead {
          background: ${token.colorFillQuaternary};
        }

        tr {
          box-shadow: 0 1px 0 var(--lobe-markdown-border-color);
        }

        th,
        td {
          min-width: 120px;
          padding-block: 0.75em;
          padding-inline: 1em;
          text-align: start;
        }
      }
    `;
    const video = css`
      > video:not(${IGNORE_CLASSNAME} video),
      > p:not(${IGNORE_CLASSNAME} p) > video {
        margin-block: calc(var(--lobe-markdown-margin-multiple) * 1em);
        border-radius: calc(var(--lobe-markdown-border-radius) * 1px);
        box-shadow: 0 0 0 1px var(--lobe-markdown-border-color);
      }

      video:not(${IGNORE_CLASSNAME} video) {
        max-width: 100%;
      }
    `;

    return cx(
      __root,
      a,
      blockquote,
      code,
      details,
      header,
      hr,
      img,
      kbd,
      list,
      p,
      pre,
      strong,
      svg,
      table,
      video,
    );
  },
);
