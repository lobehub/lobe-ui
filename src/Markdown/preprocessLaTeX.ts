export const preprocessLaTeX = (content: string) => {
  // Replace block-level LaTeX delimiters \[ \] with $$ $$

  const blockProcessedContent = content.replaceAll(
    /\\\[(.*?)\\]/gs,
    (_, equation) => `$$${equation}$$`,
  );
  // Replace inline LaTeX delimiters \( \) with $ $
  const inlineProcessedContent = blockProcessedContent.replaceAll(
    /\\\((.*?)\\\)/gs,
    (_, equation) => `$${equation}$`,
  );
  return inlineProcessedContent;
};
