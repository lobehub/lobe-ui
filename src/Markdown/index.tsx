'use client';

import type { AnchorProps } from 'antd';
import { CSSProperties, memo, useMemo } from 'react';
import ReactMarkdown from 'react-markdown';
import { Components } from 'react-markdown/lib/ast-to-react';
import rehypeKatex from 'rehype-katex';
import rehypeRaw from 'rehype-raw';
import remarkBreaks from 'remark-breaks';
import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';

import ImageGallery from '@/Image/ImageGallery';
import Image, { type ImageProps } from '@/mdx/Image';
import Link from '@/mdx/Link';
import type { PreProps } from '@/mdx/Pre';
import Video, { type VideoProps } from '@/mdx/Video';
import type { AProps } from '@/types';

import { CodeFullFeatured, CodeLite } from './CodeBlock';
import type { TypographyProps } from './Typography';
import { useStyles as useMarkdownStyles } from './markdown.style';
import { useStyles } from './style';
import { escapeBrackets, escapeMhchem } from './utils';

export interface MarkdownProps extends TypographyProps {
  allowHtml?: boolean;
  children: string;
  className?: string;
  componentProps?: {
    a?: Partial<AProps & AnchorProps>;
    img?: Partial<ImageProps>;
    pre?: Partial<PreProps>;
    video?: Partial<VideoProps>;
  };
  enableImageGallery?: boolean;
  enableLatex?: boolean;
  fullFeaturedCodeBlock?: boolean;
  onDoubleClick?: () => void;
  style?: CSSProperties;
  variant?: 'normal' | 'chat';
}

const Markdown = memo<MarkdownProps>(
  ({
    children,
    className,
    style,
    fullFeaturedCodeBlock,
    onDoubleClick,
    enableLatex = true,
    enableImageGallery = true,
    componentProps,
    allowHtml,
    fontSize,
    headerMultiple,
    marginMultiple,
    variant = 'normal',
    lineHeight,
    ...rest
  }) => {
    const { cx, styles } = useStyles({ fontSize, headerMultiple, lineHeight, marginMultiple });
    const { styles: mdStyles } = useMarkdownStyles({ fontSize, headerMultiple, marginMultiple });
    const isChatMode = variant === 'chat';

    const escapedContent = useMemo(() => {
      if (!enableLatex) return children;
      return escapeMhchem(escapeBrackets(children));
    }, [children, enableLatex]);

    const components: Components = useMemo(
      () => ({
        a: (props: any) => <Link {...props} {...componentProps?.a} />,
        img: enableImageGallery
          ? (props: any) => (
              <Image
                {...props}
                {...componentProps?.img}
                style={
                  isChatMode
                    ? { height: 'auto', maxWidth: 640, ...componentProps?.img?.style }
                    : componentProps?.img?.style
                }
              />
            )
          : undefined,
        pre: (props: any) =>
          fullFeaturedCodeBlock ? (
            <CodeFullFeatured {...props} {...componentProps?.pre} />
          ) : (
            <CodeLite {...props} {...componentProps?.pre} />
          ),
        video: (props: any) => <Video {...props} {...componentProps?.video} />,
      }),
      [componentProps, enableImageGallery, fullFeaturedCodeBlock],
    );

    const rehypePlugins = useMemo(
      () => [allowHtml && rehypeRaw, enableLatex && rehypeKatex].filter(Boolean) as any,
      [allowHtml],
    );
    const remarkPlugins = useMemo(
      () =>
        [remarkGfm, enableLatex && remarkMath, isChatMode && remarkBreaks].filter(Boolean) as any,
      [isChatMode],
    );

    return (
      <article
        className={cx(styles.root, className)}
        data-code-type="markdown"
        onDoubleClick={onDoubleClick}
        style={style}
      >
        <ImageGallery enable={enableImageGallery}>
          <ReactMarkdown
            className={cx(
              mdStyles.__root,
              mdStyles.a,
              mdStyles.blockquote,
              mdStyles.code,
              mdStyles.details,
              mdStyles.header,
              mdStyles.hr,
              mdStyles.img,
              mdStyles.kbd,
              mdStyles.list,
              mdStyles.p,
              mdStyles.pre,
              mdStyles.strong,
              mdStyles.table,
              mdStyles.video,
              isChatMode && styles.chat,
            )}
            components={components}
            rehypePlugins={rehypePlugins}
            remarkPlugins={remarkPlugins}
            {...rest}
          >
            {escapedContent}
          </ReactMarkdown>
        </ImageGallery>
      </article>
    );
  },
);

export default Markdown;
