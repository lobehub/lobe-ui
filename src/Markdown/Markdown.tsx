'use client';

import { cva } from 'class-variance-authority';
import { memo, useCallback, useMemo } from 'react';
import ReactMarkdown from 'react-markdown';
import type { Components } from 'react-markdown/lib';

import { PreviewGroup } from '@/Image';
import Image from '@/mdx/mdxComponents/Image';
import Link from '@/mdx/mdxComponents/Link';
import Section from '@/mdx/mdxComponents/Section';
import Video from '@/mdx/mdxComponents/Video';

import { CodeFullFeatured, CodeLite } from './CodeBlock';
import Typography from './Typography';
import { useStyles } from './style';
import type { MarkdownProps } from './type';
import {
  addToCache,
  contentCache,
  createPlugins,
  escapeBrackets,
  escapeMhchem,
  fixMarkdownBold,
  transformCitations,
} from './utils';

const Markdown = memo<MarkdownProps>(
  ({
    ref,
    children,
    className,
    style,
    fullFeaturedCodeBlock,
    onDoubleClick,
    enableLatex = true,
    enableMermaid = true,
    enableImageGallery = true,
    enableCustomFootnotes,
    componentProps,
    allowHtml,
    fontSize = 14,
    headerMultiple = 0.25,
    marginMultiple = 1,
    showFootnotes,
    variant = 'default',
    reactMarkdownProps,
    lineHeight = 1.6,
    rehypePlugins,
    remarkPlugins,
    remarkPluginsAhead,
    components = {},
    customRender,
    citations,
    ...rest
  }) => {
    const { cx, styles } = useStyles();
    const isChatMode = variant === 'chat';

    const variants = useMemo(
      () =>
        cva(styles.root, {
          defaultVariants: {
            enableLatex: true,
            variant: 'default',
          },
          /* eslint-disable sort-keys-fix/sort-keys-fix */
          variants: {
            variant: {
              default: null,
              chat: styles.chat,
            },
            enableLatex: {
              true: styles.latex,
              false: null,
            },
          },
          /* eslint-enable sort-keys-fix/sort-keys-fix */
        }),
      [styles],
    );

    // 计算缓存键
    const cacheKey = useMemo(() => {
      return `${children}-${enableLatex}-${enableCustomFootnotes}-${citations?.length || 0}`;
    }, [children, enableLatex, enableCustomFootnotes, citations?.length]);

    // 处理内容并利用缓存避免重复计算
    const escapedContent = useMemo(() => {
      // 尝试从缓存获取
      if (contentCache.has(cacheKey)) {
        return contentCache.get(cacheKey);
      }

      // 处理新内容
      let processedContent;
      if (enableLatex) {
        const baseContent = fixMarkdownBold(escapeMhchem(escapeBrackets(children)));
        processedContent = enableCustomFootnotes
          ? transformCitations(baseContent, citations?.length)
          : baseContent;
      } else {
        processedContent = fixMarkdownBold(children);
      }

      // 缓存处理结果
      addToCache(cacheKey, processedContent);
      return processedContent;
    }, [cacheKey, children, enableLatex, enableCustomFootnotes, citations?.length]);

    // 创建插件
    const { rehypePluginsList, remarkPluginsList } = useMemo(
      () =>
        createPlugins({
          allowHtml,
          enableCustomFootnotes,
          enableLatex,
          isChatMode,
          rehypePlugins,
          remarkPlugins,
          remarkPluginsAhead,
        }),
      [
        allowHtml,
        enableLatex,
        enableCustomFootnotes,
        isChatMode,
        rehypePlugins,
        remarkPlugins,
        remarkPluginsAhead,
      ],
    );

    // 使用 useCallback 优化渲染子组件
    const renderLink = useCallback(
      (props: any) => <Link citations={citations} {...props} {...componentProps?.a} />,
      [citations, componentProps?.a],
    );

    const renderImage = useCallback(
      (props: any) => <Image {...props} {...componentProps?.img} />,
      [isChatMode, componentProps?.img],
    );

    const renderCodeBlock = useCallback(
      (props: any) =>
        fullFeaturedCodeBlock ? (
          <CodeFullFeatured
            enableMermaid={enableMermaid}
            highlight={componentProps?.highlight}
            mermaid={componentProps?.mermaid}
            {...props}
            {...componentProps?.pre}
          />
        ) : (
          <CodeLite
            enableMermaid={enableMermaid}
            highlight={componentProps?.highlight}
            mermaid={componentProps?.mermaid}
            {...props}
            {...componentProps?.pre}
          />
        ),
      [
        enableMermaid,
        fullFeaturedCodeBlock,
        componentProps?.highlight,
        componentProps?.mermaid,
        componentProps?.pre,
      ],
    );

    const renderSection = useCallback(
      (props: any) => <Section showCitations={showFootnotes} {...props} />,
      [showFootnotes],
    );

    const renderVideo = useCallback(
      (props: any) => <Video {...props} {...componentProps?.video} />,
      [componentProps?.video],
    );

    // 创建组件映射
    const memoComponents = useMemo(
      () => ({
        a: renderLink,
        img: enableImageGallery ? renderImage : undefined,
        pre: renderCodeBlock,
        section: renderSection,
        video: renderVideo,
        ...components,
      }),
      [
        renderLink,
        renderImage,
        renderCodeBlock,
        renderSection,
        renderVideo,
        enableImageGallery,
        components,
      ],
    ) as Components;

    // 渲染默认内容
    const defaultDOM = useMemo(
      () => (
        <PreviewGroup enable={enableImageGallery}>
          <ReactMarkdown
            {...reactMarkdownProps}
            components={memoComponents}
            rehypePlugins={rehypePluginsList}
            remarkPlugins={remarkPluginsList}
          >
            {escapedContent}
          </ReactMarkdown>
        </PreviewGroup>
      ),
      [escapedContent, memoComponents, rehypePluginsList, remarkPluginsList, enableImageGallery],
    );

    // 应用自定义渲染
    const markdownContent = customRender
      ? customRender(defaultDOM, { text: escapedContent || '' })
      : defaultDOM;

    return (
      <Typography
        className={cx(variants({ enableLatex, variant }), className)}
        data-code-type="markdown"
        fontSize={fontSize}
        headerMultiple={headerMultiple}
        lineHeight={lineHeight}
        marginMultiple={marginMultiple}
        onDoubleClick={onDoubleClick}
        ref={ref}
        style={style}
        {...rest}
      >
        {markdownContent}
      </Typography>
    );
  },
);

Markdown.displayName = 'Markdown';

export default Markdown;
