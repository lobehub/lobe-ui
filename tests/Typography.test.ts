import { render } from '@testing-library/react';
import React from 'react';
import { describe, expect, it } from 'vitest';

import { Typography } from '../src/Markdown/Typography';

describe('Typography', () => {
  it('should render correctly with default props', () => {
    const { container } = render(React.createElement(Typography, null, 'Test content'));
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with custom fontSize', () => {
    const { container } = render(React.createElement(Typography, { fontSize: 16 }, 'Test content'));
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with custom headerMultiple', () => {
    const { container } = render(
      React.createElement(Typography, { headerMultiple: 1.5 }, 'Test content'),
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with custom marginMultiple', () => {
    const { container } = render(
      React.createElement(Typography, { marginMultiple: 3 }, 'Test content'),
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with custom lineHeight', () => {
    const { container } = render(
      React.createElement(Typography, { lineHeight: 2 }, 'Test content'),
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with custom className', () => {
    const { container } = render(
      React.createElement(Typography, { className: 'custom-class' }, 'Test content'),
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with multiple props', () => {
    const { container } = render(
      React.createElement(
        Typography,
        {
          className: 'custom-class',
          fontSize: 14,
          headerMultiple: 1.2,
          lineHeight: 1.5,
          marginMultiple: 2.5,
        },
        'Test content',
      ),
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with default marginMultiple when not provided', () => {
    const { container } = render(React.createElement(Typography, null, 'Test content'));
    expect(container.firstChild).toMatchSnapshot();
  });
});
