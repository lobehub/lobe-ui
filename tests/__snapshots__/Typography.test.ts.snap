// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typography > should render correctly with default props 1`] = `
<article
  class="acss-c80x8p"
>
  Test content
</article>
`;

exports[`Typography > should render with custom className 1`] = `
<article
  class="custom-class acss-c80x8p"
>
  Test content
</article>
`;

exports[`Typography > should render with custom fontSize 1`] = `
<article
  class="acss-c80x8p"
>
  Test content
</article>
`;

exports[`Typography > should render with custom headerMultiple 1`] = `
<article
  class="acss-q0drdl"
>
  Test content
</article>
`;

exports[`Typography > should render with custom lineHeight 1`] = `
<article
  class="acss-mltska"
>
  Test content
</article>
`;

exports[`Typography > should render with custom marginMultiple 1`] = `
<article
  class="acss-1g7fh7s"
>
  Test content
</article>
`;

exports[`Typography > should render with default marginMultiple when not provided 1`] = `
<article
  class="acss-c80x8p"
>
  Test content
</article>
`;

exports[`Typography > should render with multiple props 1`] = `
<article
  class="custom-class acss-1ipxt5j"
>
  Test content
</article>
`;
